truenas_api_client/utils.py

```diff
@@ -1,18 +1,45 @@
+"""Utility classes for use in the TrueNAS API client.
+
+Includes `Struct` for creating regular objects out of `Mapping`s with string
+keys and `ProgressBar` for displaying the progress of a task in the CLI.
+
+Attributes:
+    MIDDLEWARE_RUN_DIR: Directory containing the middlewared Unix domain socket.
+    undefined: A dummy object similar in purpose to `None` that indicates an unset variable.
+
+"""
 import sys
 
+from typing import Any, TextIO, Mapping
+
 
 MIDDLEWARE_RUN_DIR = '/var/run/middleware'
-undefined = object()
+undefined = -1
```

Comment:
Why?

Reply:
It's easier to stick with one type when dealing with type checkers. If `timeout` can be an `object`, its type everywhere becomes `float | object` which will cause `float` operations to be flagged since it may be an `object`, even if you check if `timeout is undefined`. Note that this change still allows checking `timeout is undefined`.

Reply:
It seems like it would be better to give `undefined` its own type then

Reply:
The specific issue I'm having is that `timeout` is `float | object` when we pass it to `wait()` which is caused by `undefined = object()`. Can we replace `undefined` with `None` and instead check `timeout is None`?

Reply:
It's not `None`, because `None` usually means "no timeout". It's a common python practice, e.g. https://github.com/pydantic/pydantic/blob/main/pydantic/fields.py#L37

Reply:
I've implemented `undefined` as a singleton. Please see if it's what we want, otherwise I'll leave it as it was.

This 3-year-old PEP tried to address the problem but never was added to the language: https://peps.python.org/pep-0661/